config/config.go

```diff
@@ -295,6 +297,9 @@ type KubernetesConfig struct {
 	// Kiali cache list of namespaces per user, this is typically short lived cache compared with the duration of the
 	// namespace cache defined by previous CacheDuration parameter
 	CacheTokenNamespaceDuration int `yaml:"cache_token_namespace_duration,omitempty"`
+	// ClusterName is the name of the kubernetes cluster that Kiali is running in.
+	// If empty, then it will default to 'Kubernetes'.
+	ClusterName string `yaml:"cluster_name,omitempty"`
```

Comment:
Yes we're requiring the user to configure this. My original plan was to pull it from the Istio ConfigMap but it's actually not present there. If you deploy istio with istioctl you can set it with one of the `values` args but those don't show up in the istio configmap. The only place I found the cluster id at runtime was on the istio deployment itself as an env var but I'm not sure if this is a canonical location for it. We could possibly query one of the istio debug endpoints for this info if istio stored this and exposed this on one of the debug routes. Rather than dig deeper into this right now, I opted for making it a config value and added a `TODO:` for auto-detecting it. Auto-detection should probably happen outside of the client factory whenever the config first gets created.

Reply:
The other scenario that needs to be accounted for is what happens when Kiali starts before Istio or if Kiali is fetching this from istiod directly, what happens when istiod is down when Kiali tries to fetch it? By having this be a config setting, users can set this manually to get around these scenarios. That being said, it'd be good if Kiali at least tried to auto-detect this setting first. I can include that in this PR if need be.

Reply:
Are we also adding the ability to the operator to configure this new value through the CR?.

Reply:
I was working on the mesh section and having multicluster installed, I saw that the name for the home cluster was "east", I checked where the value was obtained and it sounds like we've already been doing what you mentioned abou getting the value from an env var, I'm wondering if we can reuse that.

https://github.com/kiali/kiali/blob/master/business/mesh.go#L205

Reply:
So that's interesting, there's actually a lot of duplication between what's happening in that mesh service and what is happening/needs to happen in the client factory. Loading of remote secrets for example is happening in both places. IMO we should take those parts out of the mesh service and put them into the client factory.

Reply:
Yes, that mesh service, part of it, it's old and it was used by the old mc code. I agree it needs some cleanup, I don't think the secrets part is used anymore. 

I can clean that code up as part of the mesh section changes part, no problem.

Looking at this PR, I was particulary interested in the logic to get the cluster name, I think we can move it to other part of the code, but maybe we can continue doing what we were doing to obtain the name, what do you think?. At least at a first change to auto detect it, then fallback to the configuration, and lastly to the hardcoded value.

Reply:
> I can clean that code up as part of the mesh section changes part, no problem.

I think there's gonna be some overlap between the changes you're working on and the changes in this PR. I'm guessing you'll need each of the kube clients to find out all the info on the mesh? And those will probably come from the client factory? Right now there's a bit of a chicken and egg problem where the client factory needs the mesh service to determine the cluster id but then the mesh service needs the client factory to get the list of clients.

> Looking at this PR, I was particulary interested in the logic to get the cluster name, I think we can move it to other part of the code, but maybe we can continue doing what we were doing to obtain the name, what do you think?. At least at a first change to auto detect it, then fallback to the configuration, and finally to the hardcoded value.

Yeah that's what I was thinking too. I still think it'd be useful to have the config value in case of environments where istiod is running outside the cluster and Kiali can't look at the istiod deployment to determine the cluster id.

Reply:
> > I can clean that code up as part of the mesh section changes part, no problem.
> 
> I think there's gonna be some overlap between the changes you're working on and the changes in this PR. I'm guessing you'll need each of the kube clients to find out all the info on the mesh? And those will probably come from the client factory? Right now there's a bit of a chicken and egg problem where the client factory needs the mesh service to determine the cluster id but then the mesh service needs the client factory to get the list of clients.
> 

Yes, well, I'm starting to design some things on the UI, I didn't touch backend code yet, I was only reviewing what we have so far. So, if you want, we can aim to merge this first, and then I can start working on the backend side for the mesh section, is that ok for you?.

> > Looking at this PR, I was particulary interested in the logic to get the cluster name, I think we can move it to other part of the code, but maybe we can continue doing what we were doing to obtain the name, what do you think?. At least at a first change to auto detect it, then fallback to the configuration, and finally to the hardcoded value.
> 
> Yeah that's what I was thinking too. I still think it'd be useful to have the config value in case of environments where istiod is running outside the cluster and Kiali can't look at the istiod deployment to determine the cluster id.

+1

Reply:
> Yes, well, I'm starting to design some things on the UI, I didn't touch backend code yet, I was only reviewing what we have so far. So, if you want, we can aim to merge this first, and then I can start working on the backend side for the mesh section, is that ok for you?.

Yeah that sounds good